planner/core/logical_plan_builder.go

```diff
@@ -1031,6 +1031,7 @@ func (b *PlanBuilder) coalesceCommonColumns(p *LogicalJoin, leftPlan, rightPlan
 
 func (b *PlanBuilder) buildSelection(ctx context.Context, p LogicalPlan, where ast.ExprNode, aggMapper map[*ast.AggregateFuncExpr]int) (LogicalPlan, error) {
 	b.optFlag |= flagPredicatePushDown
+	b.optFlag |= flagDeriveTopN
```

Comment:
Is it better to move this operation into `buildWindowSpecs` or just check whether `len(b.windowSpecs) > 0` here? 
Since `window-function` seems rarer than `selection`, then we can let more queries that cannot fit this rule skip it.

Reply:
Got your point about limiting the rule. But, is it too late to set the flag in buildWindowSpecs since we need to check the Select parent? I think we can add a simple check here too for the window but it does not clean to me. I think it is better to encapsulate all the checks in the rule itself for readability and debugging.

Reply:
Gotit. The current implementation is OK for me.